File: app/src/main/java/gex/com/masinqojava/SongFragment.java
package gex.com.masinqojava;

import static gex.com.masinqojava.MainActivity.songs;

import android.os.Build;
import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;


public class SongFragment extends Fragment {

    RecyclerView recyclerView;
    SongAdapter songAdapter;


    public SongFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_song, container, false);
        recyclerView = view.findViewById(R.id.song_recycler);
        recyclerView.setHasFixedSize(true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (!(songs.isEmpty())){
                songAdapter = new SongAdapter(songs, getContext());
                recyclerView.setAdapter(songAdapter);
                recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false));
            }else{
                Toast.makeText(getContext(), "No songs found", Toast.LENGTH_SHORT).show();
            }
        }
        return view;
    }
}

File: app/src/main/java/gex/com/masinqojava/SongFragment.kt
package gex.com.masinqojava

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout

class SongFragment : Fragment() {
    var recyclerView: RecyclerView? = null
    var songAdapter: SongAdapter? = null

    var swipeRefreshLayout: SwipeRefreshLayout? = null

    private val newSongs by lazy { MusicLoader.getSongs(requireContext()) }
    val oldSongs = MainActivity.songs ?: emptyList()
    private val playbackViewModel: PlaybackViewModel by activityViewModels()
    private val playlistViewModel: PlaylistViewModel by activityViewModels()


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_song, container, false)
        recyclerView = view.findViewById(R.id.song_recycler)
        swipeRefreshLayout = view.findViewById(R.id.swipe_refresh)
        swipeRefreshLayout?.setOnRefreshListener {

            val diffResult = DiffUtil.calculateDiff(SongDiffCallback(oldSongs, newSongs))

            MainActivity.songs?.clear()
            MainActivity.songs?.addAll(newSongs)
            songAdapter?.let {
                diffResult.dispatchUpdatesTo(it)
            }
            swipeRefreshLayout?.isRefreshing = false
        }
        recyclerView!!.setHasFixedSize(true)

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val songs = MainActivity.songs
        if (!songs.isNullOrEmpty()){
            songAdapter = SongAdapter(
                songs = songs,
                context = requireContext(),
                viewModel = playbackViewModel,
                onItemClick = {pos->
                    playbackViewModel.playList(songs,pos)
                },
                onMoreOptionClick = {song ->
                    val optionSheet = SongOptionBottomSheet(song, playbackViewModel,playlistViewModel)
                    optionSheet.show(childFragmentManager, "SongOptions")
                }
            )
            recyclerView?.adapter = songAdapter
            recyclerView?.layoutManager = LinearLayoutManager(requireContext())
        }else{
            Toast.makeText(context, "No songs found", Toast.LENGTH_SHORT).show()
        }
    }

    fun filterSongs(query: String) {
        songAdapter?.filter?.filter(query)
    }

}